src/main/java/com/cloudbees/jenkins/support/filter/FilteredWriter.java

```diff
@@ -0,0 +1,138 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.cloudbees.jenkins.support.filter;
+
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.GuardedBy;
+import java.io.FilterWriter;
+import java.io.IOException;
+import java.io.Writer;
+import java.nio.CharBuffer;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Wraps a Writer by filtering written lines using a provided ContentFilter.
+ *
+ * @see ContentFilter
+ * @see FilteredOutputStream
+ * @since TODO
+ */
+public class FilteredWriter extends FilterWriter {
+
+    static final Pattern EOL = Pattern.compile("\r?\n");
+    static final int DEFAULT_BUFFER_CAPACITY = 1024;
+
+    private final ContentFilter contentFilter;
+    @GuardedBy("this")
+    private CharBuffer buf = CharBuffer.allocate(DEFAULT_BUFFER_CAPACITY);
+
+    public FilteredWriter(@Nonnull Writer writer, @Nonnull ContentFilter contentFilter) {
+        super(writer);
+        this.contentFilter = contentFilter;
+    }
+
+    @Override
+    public synchronized void write(int c) throws IOException {
+        write(Character.toChars(c));
+    }
+
+    @Override
+    public synchronized void write(@Nonnull char[] cbuf, int off, int len) throws IOException {
+        while (len > 0) {
+            if (!buf.hasRemaining()) filterFlushLines();
+            if (!buf.hasRemaining()) {
+                buf = CharBuffer.allocate(buf.capacity() + DEFAULT_BUFFER_CAPACITY).put(buf);
+            }
+            int toCopy = Math.min(buf.remaining(), len);
+            if (toCopy == 0) throw new IllegalStateException();
+            buf.put(cbuf, off, toCopy);
+            filterFlushLines();
+            len -= toCopy;
+            off += toCopy;
+        }
+        filterFlushLines();
+    }
+
+    @Override
+    public synchronized void write(@Nonnull String str, int off, int len) throws IOException {
+        while (len > 0) {
+            if (!buf.hasRemaining()) filterFlushLines();
+            if (!buf.hasRemaining()) {
+                buf = CharBuffer.allocate(buf.capacity() + DEFAULT_BUFFER_CAPACITY).put(buf);
+            }
+            int toCopy = Math.min(buf.remaining(), len);
+            if (toCopy == 0) throw new IllegalStateException();
+            buf.put(str, off, off + toCopy);
+            filterFlushLines();
+            len -= toCopy;
+            off += toCopy;
+        }
+        filterFlushLines();
+    }
+
+    @Override
+    public synchronized void flush() throws IOException {
+        if (buf.position() > 0) {
+            buf.flip();
+            String original = buf.toString();
+            String filtered = contentFilter.filter(original);
+            out.write(filtered);
+            buf.clear();
+            if (buf.capacity() > DEFAULT_BUFFER_CAPACITY) {
+                buf = CharBuffer.allocate(DEFAULT_BUFFER_CAPACITY);
```

Comment:
I would not resize the buffer here, see my earlier comment. Also, if you reused the buffer from the parent class couldn't you use its implementation as well?